Gem/data uploader develop

agnessa opened the pull request:
Data uploader logic packaged into a gem. To test this, please use this branch of Climate Watch:
https://github.com/Vizzuality/climate-watch/pull/707

Depending on whether you have previously worked on data uploader, you might need to:
- `bundle exec rake db:migrate`
- `bundle exec rake db:admin_boilerplate_create`

Remember to start sidekiq before testing.

Have a look at the README of the data uploader gem, it explains the steps needed to integrate the gem in host.

I moved hopefully all data upload related stuff from Climate Watch to here. This includes:
- rake tasks
- AR models (dataset, section ...) and their migrations
- admin utility classes
- workers

There are very few specs, because it took me ages to get even those simple engine specs to work and I ran out of time 😓 

Notable things which changed:
- I thought it best to wrap all the classes from the gem in the same namespace. This way, when used in host applications it is clear they are external and come from the gem. The namespace is DataUploader and everything (models, utility classes, workers etc) should sit under it; in the process I removed the Admin namespace from some of the classes that had it.
- the config/database_uploader.yml file has 2 functions: it is used by the `data_uploader:generate` rake task to generate controllers, and by the `db:admin_boilerplate_create` task to generate the AR objects.
- `APP_NAME` became `APP_NAMES` and is a comma-separate list of platforms; I'm not sure if this is going against a different design, but I think at some point we thought global, India & Indonesia could be editable from the same admin panel. Let me know if this is completely wrong.

Ideally let's merge https://github.com/ClimateWatch-Vizzuality/climate-watch-gems/pull/5 first, so that we can remove the dependency on climate_watch_engine from that branch.

aabdaab commented on data_uploader/app/admin/data_uploader/use_case/download_zipped_datafiles.rb:
```diff
@@ -0,0 +1,40 @@
+
+        zip_filename = "#{attrs[:platform_name]}-#{attrs[:section_name]}-datafiles"
+
+        Admin::ZipAndDownload.call(
```
same here: `DataUploader::ZipAndDownload`

aabdaab commented on data_uploader/app/admin/data_uploader/use_case/upload_datafile.rb:
```diff
@@ -0,0 +1,44 @@
+
+        dataset.datafile.attach(params[:datafile])
+
+        Admin::S3Uploader.call(
```
and here: `DataUploader::S3Uploader`

agnessa commented:
Thanks @aabdaab, fixed those namespace issues, not sure how I was testing this - sorry for the mess.

aabdaab commented on data_uploader/app/admin/data_uploader/use_case/upload_datafile.rb:
```diff
@@ -0,0 +1,44 @@
+  module UseCase
+    class UploadDatafile < DataUploader::UseCase::Base
+      def call(params, callbacks)
+        validate_if_file_chosen params
```
just realized that the validations crash the app; we need to pass `callbacks` to the validation methods. Both to `validate_if_file_chosen ` and `validate_content_type_against_csv`. Sorry for that, I must've missed that

aabdaab commented:
Is it normal that worker logging is off (because of local testing?)?

aabdaab commented:
Not sure about the `APP_NAMES`. That was the initial idea but along the way we decided it'd be easier if we had the view only for the app we're actually on. So, when on Climate Watch, we allow uploading and importing only for CW. While the uploading part would be pretty easy to do it for the other platforms, running the importers wouldn't be as straightforward, as it would have to do the import to the project's database. The same for keeping the logs in sync; if we would want to import datafiles in South Africa platform for Climate Watch we wouldn't have the logs from previous workers, as they sit in the CW database. I thought about putting them in the shared database but they're in relationship with section so that would have to go to the shared database as well ---> which I thought about all along too, but ActiveStorage works only with the default database.

agnessa commented:
> Not sure about the `APP_NAMES`. That was the initial idea but along the way we decided it'd be easier if we had the view only for the app we're actually on. 

OK, will revert!

agnessa commented:
> Is it normal that worker logging is off (because of local testing?)?

Don't think it's off - output goes to log/sidekiq.log though.

agnessa commented:
I addressed the comments I think and bumped the version of the gem so that it hopefully installs correctly in https://github.com/Vizzuality/climate-watch/pull/707

aabdaab commented:
Two things are not working: 
When hitting upload when no file was chosen:
![image](https://user-images.githubusercontent.com/6136899/47359257-2fe0dd00-d6c4-11e8-907e-07d6b654f3a4.png)
It shouldn't reach the content_type validation as it should be returned from the actions in here `validate_if_file_chosen params, callbacks` not sure, why it's not, I looked at the code and it seems fine 🤔 
And second thing, uploading other file than csv should show an error that csv file is require, instead it crashes the page: 
![image](https://user-images.githubusercontent.com/6136899/47359472-c7dec680-d6c4-11e8-9fd9-ee41428094b1.png)
This should also be handled in `validate_content_type_against_csv params, callbacks`

Another thing I noticed is that the NDC Texts disappeared from the dropdown. This one was acting differently than the other pages, it accepted only one dataset which was supposed to be zipped file and then we unzipped it and uploaded to S3. It's present in this PR: https://github.com/Vizzuality/climate-watch/pull/703/files#diff-c5a621af4be29ea1a00be53fe4607556R1 
The problem with that is that it's specific for CW platform but maybe it is still worth it to abstract this as this could be anything that we upload as zipped files.

agnessa commented:
> It shouldn't reach the content_type validation as it should be returned from the actions in here validate_if_file_chosen params, callbacks

So this breaks my brain every time, but apparently if we need to terminate execution in the caller, we need to use a proc instead of a lambda 😅 Now with a proc it exits after a failed validation.

agnessa commented:
> Another thing I noticed is that the NDC Texts disappeared from the dropdown.

I'm leaving it as it was originally (except there were some namespace changes that needed amending), it shows up for me in the dropdown.

agnessa commented:
> The problem with that is that it's specific for CW platform but maybe it is still worth it to abstract this as this could be anything that we upload as zipped files.

Maybe let's abstract when we have a use case, can't see one for now.
